### PR TITLE
[docs] Add advanced tutorials for SEO and social sharing cards

### LINKED ISSUE #14125: [docs] Add more details on implementing SEO techniques in Gatsby
## Summary

In PR #10780, @DSchau proposed a really great piece of content on the blog on implementing SEO techniques in Gatsby. We decided to close that PR in favor of a new issue for a Reference Guide instead of a blog post–but the content from the PR could be copied and slightly modified. I think it makes the most sense to add it as an Advanced Tutorial (dependent on #14071) and link to that from the [SEO docs page](https://www.gatsbyjs.org/docs/seo/).

To fit in with tutorials and docs, any mentions of "us" or "we" should change to "you" to directly address the reader (more in the [Gatsby Style Guide](https://www.gatsbyjs.org/contributing/gatsby-style-guide/#use-you-as-the-pronoun))

This will be a great how-to for adding SEO stuff to Gatsby sites, especially explaining what's in some of the starters. As a tutorial instead of a blog post, it will be easier to find after more content is added since it will be more discoverable through the docs.

### PR DESCRIPTION
**Pull request by FlorianDr**
## Description
This is about moving the initial blog post by @DSchau into a new advanced tutorial and link it in the reference guide on the SEO page.

## Related Issues
Fixes #14125

### CODE REVIEW

**Comment by DSchau** [comment]
Cool 😎 Thanks for doing this!

I'll defer to @gatsbyjs/learning for the content review 👍

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+  }),
+  meta: PropTypes.array,
+  title: PropTypes.string.isRequired,
+}
+// highlight-end
+
+export default SEO
+```
+
+Woo hoo! You enabled not only SEO for search engines like Google, Bing (people use Bing, right?) but you also laid the groundwork for enhanced sharing capabilities on social networks. Finally, you will learn to add support for one of the more useful functionalities for SEO, specifically a canonical link.
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
Woo hoo! You enabled not only SEO for search engines like Google and Bing, but you also laid the groundwork for enhanced sharing capabilities on social networks. Finally, you will learn to add support for one of the more useful functionalities for SEO, specifically a canonical link.
```

I feel like the side comment makes more sense in a blog post but not so much a tutorial.

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+
+In addition to SEO for actual _search_ engines you also want those pretty cards that social networks like Twitter and Slack enable. Specifically, the implementation should feature:
+
+- Description for embedded results
+- Title for embedded results
+- (Optionally) display an image and a card if an image is passed in to the component
+
+Let's implement it 👌
+
+```jsx:title=src/components/seo.js
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
```js:title=src/components/seo.js
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+
+### Creating the SEO component
+
+Using the power and flexibility of React, you can create a React component to power this functionality.
+
+> Note: `react-helmet` is enabled, by default, in gatsby-starter-default and gatsby-starter-blog
+>
+> If you're not using those starters [follow this guide for installation instructions][gatsby-plugin-react-helmet]
+
+```jsx:title=src/components/seo.js
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
```js:title=src/components/seo.js
```

There are several of these that I'd like to make more consistent. I'll add suggestions to adjust them throughout.

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+}
+
+export default SEO
+```
+
+This component doesn't _do_ anything yet, but it's the foundation for a useful, extensible component. It leverages the `StaticQuery` functionality enabled via Gatsby to query siteMetadata (e.g. details in `gatsby-config.js`) with description and keywords.
+
+The `StaticQuery` component accepts a render prop, and at this point, it simply returns `null` to render nothing. Let's _actually_ render something and build out the prototype for this SEO component. Let's iterate further.
+
+```jsx:title=src/components/seo.js
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
```js:title=src/components/seo.js
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+A canonical link is a hint to a search engine that this is the _source_ for this content. It helps resolve duplicate content issues. For instance, if you have several paths to the same content, you can use a canonical link as akin to a soft redirect which will **not** harm your search ranking if implemented correctly.
+
+To implement this functionality, you need to do the following:
+
+1. Enable passing a `pathname` prop to your SEO component
+1. Prefix your `pathname` prop with your `siteUrl` (from `gatsby-config.js`)
+   - A canonical link should be _absolute_ (e.g. https://your-site.com/canonical-link), so you will need to prefix with this `siteUrl`
+1. Tie into the `link` prop of `react-helmet` to create a `<link rel="canonical" >` tag
+
+```jsx
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
```js:title=src/components/seo.js
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+
+To bring it all home, it's high time to begin actually _using_ this extensible SEO component to show all of these moving parts coming together to deliver a great SEO experience.
+
+## Using the SEO component
+
+You created an extensible SEO component. It takes a `title` prop, and then (optionally) `description`, `meta`, `image`, and `pathname` props. Let's wire it all up!
+
+### In a page component
+
+```jsx:title=src/pages/index.js
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
```js:title=src/pages/index.js
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+
+Feel free to add whatever, or perhaps use this _very pertinent_ image:
+
+![Sample Image](./images/seo.jpg)
+
+the image will need to be located at `content/blog/2019-01-04-hello-world-seo/images/featured.jpg`
+
+#### Querying with GraphQL
+
+```jsx:title=src/templates/blog-post.js
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
```js:title=src/templates/blog-post.js
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+
+```shell
+npx gatsby new my-new-gatsby-app
+```
+
+and you'll have the `SEO` component available to maximize your SEO and social sharing capabilities. Check it out!
+
+## Further Learning
+
+This tutorial is merely a shallow dive into the depths of SEO optimization--consider it a primer for further learning and a gentle introduction to some SEO concepts with a Gatsby twist. To truly master these concepts is outside the scope of this tutorial, but it truly is fascinating stuff that can directly lead to more eyes on your content!
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
This tutorial is merely a shallow dive into the depths of SEO. Consider it a primer for further learning and a gentle introduction to some SEO concepts with a Gatsby twist. To truly master these concepts is outside the scope of this tutorial, but it truly is fascinating stuff that can directly lead to more eyes on your content!
```

"search engine optimization optimization" 😄

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md (hunk 1 above)
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
Woo hoo! You enabled not only SEO for search engines like Google, Bing (people use Bing, right?) but you also laid the groundwork for enhanced sharing capabilities on social networks. Finally, you will learn to add support for one of the more useful functionalities for SEO: a canonical link.
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+  meta: PropTypes.array,
+  // highlight-next-line
+  slug: PropTypes.string,
+  title: PropTypes.string.isRequired,
+}
+
+export default SEO
+```
+
+Woo hoo! Lot to digest here, but you've enabled adding an _absolute_ canonical link simply by passing in a `slug` prop and prefixing with `siteUrl`.
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
Woo hoo! Lots to digest here, but you've enabled adding an _absolute_ canonical link by passing in a `slug` prop and prefixing with `siteUrl`.
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+
+To bring it all home, it's high time to begin actually _using_ this extensible SEO component to show all of these moving parts coming together to deliver a great SEO experience.
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
To bring it all home, it's time to begin actually _using_ this extensible SEO component to show all of these moving parts coming together to deliver a great SEO experience.
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+
+## Using the SEO component
+
+You created an extensible SEO component. It takes a `title` prop, and then (optionally) `description`, `meta`, `image`, and `pathname` props. Let's wire it all up!
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
You created an extensible SEO component. It takes a `title` prop and then (optionally) `description`, `meta`, `image`, and `pathname` props. Let's wire it all up!
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+---
+
+Hello World!
+```
+
+#### Adding the image
+
+Feel free to add whatever, or perhaps use this _very pertinent_ image:
+
+![Sample Image](./images/seo.jpg)
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
![Sample Image](./images/seo.jpg)
```

If you choose to maintain the sentence immediately preceding this image, I suggest making the alt text more specific. Anyone who can't see the image won't get the joke!

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 13]
```diff
@@ -0,0 +1,599 @@
...
+
+the image will need to be located at `content/blog/2019-01-04-hello-world-seo/images/featured.jpg`
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
The image will need to be located at `content/blog/2019-01-04-hello-world-seo/images/featured.jpg`.
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+![Facebook](./images/facebook.png)
+
+### Twitter
+
+![Twitter](./images/twitter.png)
+
+### Slack
+
+![Slack](./images/slack.png)
+
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]: Please provide more specific alt text for these images. As is, it'll read as something like "Image: Google" and so on, which is not super descriptive.

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+To learn more about these validations, check out how to _validate_ SEO with the following tools from [Google][google-validation], [Twitter][twitter-validation], and [Facebook][facebook-validation].
+
+These SEO resources outlined in this tutorial aren't _only_ a best practice, they're also a best practice enabled, by default. Available **today** in `gatsby-starter-default`, simply use:
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
The SEO resources outlined in this tutorial aren't _only_ a best practice, they're also a best practice enabled by default. Available **today** in `gatsby-starter-default`, use:
```

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 16]
```diff
@@ -0,0 +1,599 @@
...
+
+```shell
+npx gatsby new my-new-gatsby-app
```
Comment (AishaBlake) [inline docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]:
```suggestion
npm gatsby new my-new-gatsby-app
```

This is mostly for consistency with other docs.

File: docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md [hunk 17]
```diff
@@ -0,0 +1,599 @@
...
+
+Perhaps you've even been approached by an SEO _expert_ who can maximize your revenue and page views just by following these **Three Simple Tricks**! However, relatively few make the concerted effort to implement SEO in their web app. This tutorial will share some of the ins and outs of SEO and how you can implement common, simple SEO patterns in your Gatsby web app, today. By the end of this post you'll know how to do the following:
+
+- Implement SEO patterns with [react-helmet][react-helmet]
+- Create an optimized social sharing card for Twitter, Facebook, and Slack
+- Tweak the SEO component exposed in the default gatsby starter ([`gatsby-starter-default`][gatsby-starter-default])
+
+## Implementation
+
+The core technology powering SEO is the humble, ubiquitiuous `meta` tag along with common tags like `title`. You've probably seen something like the below:
```
[the remark this answers is not in the document]
Comment (FlorianDr) [reply docs/tutorial/seo-and-social-sharing-cards-tutorial/index.md]: Thank you for the pointer. I will have a look at the feedback and try to incorporate it.

**Comment by AishaBlake** [comment]
Hi, @FlorianDr! Checking in on this and to see if there's anything I can help with or clarify.

**Comment by FlorianDr** [comment]
@AishaBlake thank you for offering your help. I started implementing your feedback, but I haven't brought it to next review yet. I will keep you posted about it. Let me know, if it's urgent to finish it ASAP.

**Comment by FlorianDr** [comment]
@AishaBlake I would be keen to hear your feedback again :) Sorry for the delay with this matter.

**Comment by AishaBlake** [comment]
Thanks, @FlorianDr! I made some quick edits but realized I missed a larger update. The code samples should be updated to reflect the current versions. If we want folks to be able to follow along, we ought to start the tutorial off at `seo.j`'s starting point in the default starter. We now use the `useStaticQuery` hook and give devs more meta tags out of the box than is indicated here. It would also be good to point out the [compatibility notes](http://localhost:8000/packages/gatsby-plugin-react-helmet/#compatibility-with-react-168-useeffect-hook) on `gatsby-plugin-react-helmet`'s page.

**Comment by AishaBlake** [comment]
The last thing I think this needs is to close the loop and show how'd you'd add a featured image, then what _that_ would look like when linked on social media. This feels outside the scope of this PR, though, so I'll make an issue for it!

**Comment by AishaBlake** [comment]
I'm actually going to go ahead and merge to keep this moving. Thank you so, so much for contributing, @FlorianDr. Really happy to finally get this content somewhere live!

**Comment by FlorianDr** [comment]
@AishaBlake Thank you so much for helping me getting this over the finish line. I learned so much.

**Pull request merged**